File: app/src/main/java/com/fstyle/structure_android/screen/main/MainPresenter.java
package com.fstyle.structure_android.screen.main;

import com.fstyle.structure_android.data.model.User;
import com.fstyle.structure_android.data.repository.UserRepository;
import com.fstyle.structure_android.data.source.remote.config.error.BaseException;
import com.fstyle.structure_android.utils.common.StringUtils;
import com.fstyle.structure_android.utils.rx.BaseSchedulerProvider;
import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

/**
 * Created by le.quang.dao on 10/03/2017.
 */

class MainPresenter implements MainContract.Presenter {

    private MainContract.View mMainView;
    private UserRepository mUserRepository;
    private CompositeDisposable mCompositeDisposable = new CompositeDisposable();
    private BaseSchedulerProvider mSchedulerProvider;

    MainPresenter(UserRepository userRepository) {
        mUserRepository = userRepository;
    }

    @Override
    public void setSchedulerProvider(BaseSchedulerProvider schedulerProvider) {
        mSchedulerProvider = schedulerProvider;
    }

    @Override
    public void setView(MainContract.View view) {
        mMainView = view;
    }

    @Override
    public void onStart() {

    }

    @Override
    public void onStop() {
        mCompositeDisposable.clear();
    }

    @Override
    public void validateDataInput() {
        boolean isValid = validateKeywordInput(mMainView.getKeyword()) & validateLimitNumberInput(
                mMainView.getLimitNumber());
        if (isValid) {
            mMainView.onDataValid();
        }
    }

    @Override
    public void searchUsers() {
        Disposable subscription =
                mUserRepository.getAllUser()
                        .subscribeOn(mSchedulerProvider.io())
                        .observeOn(mSchedulerProvider.ui())
                        .subscribe(new Consumer<List<User>>() {
                            @Override
                            public void accept(List<User> users) {
                                mMainView.onSearchUsersSuccess(users);
                            }
                        }, new Consumer<Throwable>() {
                            @Override
                            public void accept(Throwable throwable) throws Exception {
                                mMainView.onSearchError((BaseException) throwable);
                            }
                        });
        mCompositeDisposable.add(subscription);
    }

    private boolean validateKeywordInput(String keyword) {
        if (StringUtils.isBlank(keyword)) {
            mMainView.onInvalidKeyWord("Keyword must not empty!");
        }
        return !StringUtils.isBlank(keyword);
    }

    private boolean validateLimitNumberInput(String limit) {
        if (StringUtils.isBlank(limit)) {
            mMainView.onInvalidLimitNumber("Limit number must not empty!");
        }
        return !StringUtils.isBlank(limit);
    }
}

File: app/src/main/java/com/fstyle/structure_android/screen/main/MainPresenter.kt
package com.fstyle.structure_android.screen.main

import android.util.Log
import com.fstyle.structure_android.data.source.UserRepository
import com.fstyle.structure_android.data.source.remote.api.error.BaseException
import com.fstyle.structure_android.screen.BaseViewModel
import com.fstyle.structure_android.utils.common.StringUtils
import com.fstyle.structure_android.utils.rx.BaseSchedulerProvider
import com.fstyle.structure_android.utils.validator.Validator
import io.reactivex.disposables.CompositeDisposable

/**
 * Created by le.quang.dao on 10/03/2017.
 */

internal class MainPresenter(private val mUserRepository: UserRepository,
    private val mValidator: Validator) : MainContract.Presenter {

  private lateinit var mSchedulerProvider: BaseSchedulerProvider
  private lateinit var mMainViewModel: MainContract.ViewModel
  private val mCompositeSubscription: CompositeDisposable = CompositeDisposable()

  init {
    val subscription = mValidator.initNGWordPattern()
    mCompositeSubscription.add(subscription)
  }

  override fun onStart() {}

  override fun setSchedulerProvider(schedulerProvider: BaseSchedulerProvider) {
    mSchedulerProvider = schedulerProvider
  }

  override fun onStop() {
    mCompositeSubscription.clear()
  }

  override fun setViewModel(viewModel: MainContract.ViewModel) {
    mMainViewModel = viewModel
  }

  override fun validateKeywordInput(keyword: String?) {
    var message: String? = mValidator.validateValueNonEmpty(keyword)
    if (StringUtils.isBlank(message)) {
      message = mValidator.validateNGWord(keyword)
    }
    mMainViewModel.onInvalidKeyWord(message)
  }

  override fun validateLimitNumberInput(limit: String?) {
    var message: String? = mValidator.validateValueNonEmpty(limit)
    if (StringUtils.isBlank(message)) {
      message = mValidator.validateValueRangeFrom0to100(limit)
    }
    mMainViewModel.onInvalidLimitNumber(message)
  }

  override fun validateDataInput(keyword: String?, limit: String?): Boolean {
    validateKeywordInput(keyword)
    validateLimitNumberInput(limit)
    try {
      return mValidator.validateAll<BaseViewModel>(mMainViewModel)
    } catch (e: IllegalAccessException) {
      Log.e(TAG, "validateDataInput: ", e)
      return false
    }

  }

  override fun searchUsers(keyWord: String?, limit: Int) {
    val subscription = mUserRepository.searchUsers(keyWord, limit)
        .subscribeOn(mSchedulerProvider!!.io())
        .doOnSubscribe { mMainViewModel.onShowProgressBar() }
        .observeOn(mSchedulerProvider!!.ui())
        .doAfterTerminate { mMainViewModel.onHideProgressBar() }
        .subscribe({ users -> mMainViewModel.onSearchUsersSuccess(users) },
            { error -> mMainViewModel.onRequestServerError(error as BaseException) })
    mCompositeSubscription.add(subscription)
  }

  companion object {
    private val TAG = MainPresenter::class.java.name
  }
}
